### ISSUE TITLE
Incubation: Scoped Focusgroup

### ISSUE THREAD

**Issue by janewman** [w3ctag/design-reviews#1152]
### Explainer

https://open-ui.org/components/scoped-focusgroup.explainer/

### The explainer

- [x] Includes the information requested by the [Explainer Explainer](https://w3ctag.github.io/explainer-explainer/#introduction).
- [x] Follows the [Web Platform Design Principles](https://www.w3.org/TR/design-principles/).
- [x] Includes or links to answers to the [Security/Privacy Questionnaire](https://www.w3.org/TR/security-privacy-questionnaire/).
- [ ] Describes user research you did to validate the problem and/or design.

### Where and by whom is the work is being done?

- GitHub repo: https://github.com/openui/open-ui
- Primary contacts: 
  - Jacques Newman (@janewman), Microsoft, Explainer author, Chromium Implementor
  - Chris Holt (@chrisdholt), Microsoft, Prior Explainer co-author
  - Leo Lee (@[leolee](https://github.com/leotlee)), Microsoft, team lead
- Organization/project driving the design: Microsoft
- This work is being funded by: Microsoft
- Incubation and standards groups that have discussed the design:
  - [Open UI minutes from discussion](https://www.w3.org/2025/09/04-openui-minutes.html#b104)
  - [Aria WG minutes from discussion](https://www.w3.org/2025/08/21-aria-minutes.html#97da)
- Standards group(s) that you expect to discuss and/or adopt this work when it's
  ready: [WhatWG](https://github.com/whatwg/html/issues/11641)

### Feedback so far

- Multi-stakeholder feedback:
  - [Chromium is Prototyping](https://chromestatus.com/feature/5637601087193088)
  - [Mozilla comments](https://github.com/mozilla/standards-positions/issues/631)
  - [WebKit comments](https://github.com/WebKit/standards-positions/issues/171)
- Major unresolved issues with or opposition to this design:

### You should also know that...

An earlier version of this feature was proposed and resolved here: https://github.com/w3ctag/design-reviews/issues/732

---

Track conversations at https://tag-github-bot.w3.org/gh/w3ctag/design-reviews/1152

**Comment by janewman** [comment]
Thanks @matatk, these are great questions, my responses to each are below.
**_1. Orientation of certain types of construct, and aria-orientation don't seem to be mentioned - is orientation something the group considered?_**
My thinking has been that focusgroup should not be used as an input for aria-orientation. Today, the aria-orientation can be derived from the role, and focusgroup plays nicely here with the concept of a minimum role. By default, a *linear focusgroup uses both of the arrow-key axis to enable movement between items, though authors may decide to restrict this to a single axis to enable the other axis to be used for other purposes, such as scrolling, activating items, etc. Since focusgroup will never impact the actual orientation of a control, I have felt that authors should continue to use aria-orientation as they would today, and take care to either use the default, or ensure they restrict to the proper axis for the behavior or role they have chosen.
_*In the current proposal, all of the behaviors listed are linear.

**_2. Is there a reason why tree and treegrid controls are not mentioned? (grid is mentioned as future work)_**
Grid was specifically called out as future work as a previous version of the proposal had spent considerable time detailing the behavior, though this latest proposal moved it out into this future section due to complexity. 
Tree and treegrid controls would both make sense as future non-linear "behaviors" alongside grid that focusgroup could be used to support in the future.

**_3. Do you intend that the behavior tokens will always have the same names as their ARIA role counterparts?_**
My intention when adding the concept of behavior was to ensure that a focusgroup would be accessible-by-default, and to choose names that are already familiar and clear to developers. I wouldn't be against behaviors not sharing a name with the ARIA role it maps to, as long as the usage was clear to developers and accessible-by-default. 
As a **hypothetical** example, when considering how [treegrid](https://www.w3.org/WAI/ARIA/apg/patterns/treegrid/examples/treegrid-1/) could be implemented, if there was consensus that the cell vs row focus priority would be best captured by the behavior name, we might end up with "row-treegrid", "cell-treegrid", where both behaviors would map to the same role. For clarity, I might advocate that this distinction should be captured in an optional modifier, but this works as an example of the kinds of situations where there might be an interest to not have the behavior share a name with an ARIA role that it maps to.

**_4. Do you foresee adding anything that is not currently covered by ARIA?_**
When I consider the future of focusgroup, I see it following and helping to codify these common patterns, and as I see it, ARIA has a similar goal. 
While I can imagine a situation where a new role/pattern is codified in ARIA and then implemented by a corresponding focusgroup behavior, I find it difficult to imagine a situation where the reverse could be true while still maintaining a default-good accessible experience.

__________________
Please let me know if anything isn't clear, if you have additional questions, or if there are any concerns.
Thanks!

**Comment by matatk** [comment]
Thanks for your clear and comprehensive reply, @janewman - this sounds good.

Just one point I'd like to clarify regarding question 1, about orientation. I think I wasn't clear in my question (sorry about that), and that you've answered "will focusgroup affect the exposed `aria-orientation` attribute value?" - but I was meaning to ask "will focusgroup be affected by the given `aria-orientation` attribute value?"

Your answer, plus your reminders of the relevant parts of the explainer, do answer my question though - by default, both axes of arrow keys will respond, and the author can limit that to the `inline` or `block` axis as desired. It also seems that the value of `aria-orientation` is _not_ influential in any way (though the `role` of controls may be). If I'm understanding correctly, this does mean that there may be some cases where the author needs to specify both an `aria-orientation` value, and an optional `inline` or `block` constraint for the `focusgroup`, but that doesn't seem too onerous (and is consistent with the rest of how `focusgroup` works).

tl;dr: Again this all sounds reasonable. It will be great to see this feature come to the web!

**Comment by janewman** [comment]
That all lines up with my thinking, thanks for reviewing!

**Comment by matatk** [comment]
@janewman: We see that there isn't an official position from Gecko or WebKit. Are you aware of any implementation concerns that they may have, or of any forthcoming commitment to implement?

**Comment by janewman** [comment]
Mozilla has had some comments on the Draft HTML Spec I've authored, see: https://github.com/whatwg/html/pull/11723, but I'm still waiting to get feedback from Apple.
Focusgroup has been discussed at length with people working on the other browser engines in several forums at this point, AriaWG, OpenUI, and WHATWG, and I've updated the explainer and chromium implementation to address any concerns brought up in those discussions.

I'm not aware of any commitments to implement.

**Comment by janewman** [comment]
A couple of updates:
* Mozilla has moved their position to positive in https://github.com/mozilla/standards-positions/issues/1348
* Webkit has raised concerns about device independence, these concerns should be addressed with recent WPT changes, see [my recent comment on the standards position](https://github.com/WebKit/standards-positions/issues/171#issuecomment-4253966327).

**Comment by janewman** [comment]
Another update:
* focusgroup will be shipping in chromium browsers in the 150 release, see: https://chromestatus.com/feature/5637601087193088

**Comment by matatk** [comment]
Hi @janewman, thanks for your latest updates. The question about modifier keys in particular is a good one! It seems that you've addressed the concerns, and haven't had any pushback. We have a few more thoughts/questions...

1. Please could you update the explainer with a mention of (or even just a link to) [the considerations you added around modifier keys](https://github.com/whatwg/html/pull/11723/changes/1017afbf2920147b4658a986d71813936e431a57)? (We think such a mention would assist future reviewers.)

2. There was a question about spatial navigation that has been raised in community discussion, and we're not sure if it's been answered: some systems provide a spatial navigation mode (e.g. via D-pad, or arrow keys). Some of those may be due to the type of device (e.g. TV), or maybe it's a way to provide a different means of navigation as an assistive technology. How might that fit in with this proposal?

   - Specifically: [a question was raised about how this may interact with the proposed CSS Spatial Navigation](https://github.com/whatwg/html/issues/11641#issuecomment-4028039082)

**Comment by janewman** [comment]
> Hi [@janewman](https://github.com/janewman), thanks for your latest updates. The question about modifier keys in particular is a good one! It seems that you've addressed the concerns, and haven't had any pushback. We have a few more thoughts/questions...
> 
> 1. Please could you update the explainer with a mention of (or even just a link to) [the considerations you added around modifier keys](https://github.com/whatwg/html/pull/11723/changes/1017afbf2920147b4658a986d71813936e431a57)? (We think such a mention would assist future reviewers.)
> 2. There was a question about spatial navigation that has been raised in community discussion, and we're not sure if it's been answered: some systems provide a spatial navigation mode (e.g. via D-pad, or arrow keys). Some of those may be due to the type of device (e.g. TV), or maybe it's a way to provide a different means of navigation as an assistive technology. How might that fit in with this proposal?
>    
>    * Specifically: [a question was raised about how this may interact with the proposed CSS Spatial Navigation](https://github.com/whatwg/html/issues/11641#issuecomment-4028039082)

I've added the relevant sections to the explainer, see: [Modifier Keys](https://open-ui.org/components/scoped-focusgroup.explainer/#modifier-keys) and [CSS spatial navigation](https://open-ui.org/components/scoped-focusgroup.explainer/#css-spatial-navigation)

Regarding how these features work together, outside of what I covered in the explainer, I thought that spatial navigation and focusgroup compose quite well when used together today. On a site with lots of interactive elements powered by focusgroup, spatial navigation managed focus across the page, while focusgroup took over for each control. In my opinion, this gives a good mix of UA +user defined movement for navigating across the page, while giving authors the ability to control the focus behavior of their interactive controls. 
(I used this [demo site I put together](https://microsoftedge.github.io/Demos/focusgroup/index.html) to test)

When I imagine the future of keyboard focus navigation on the web, I see spatial navigation being a setting a user or user agent can turn on, with nice customization to suite the user depending on what device they are using, and with focusgroup being a more controlled interaction, where authors and users can rely on a consistent behavior. This aligns with the expectation that when spatial navigation is enabled, a user can use the keyboard to navigate on any page, while focusgroup is more about authors being able to provide a consistent focus behavior on their controls, without needing JavaScript (except for selection/activation of course).

**Comment by lukewarlow** [comment]
Hi @janewman, thanks for addressing those requests in the explainer.

We've discussed this further and would like to wait for a formal position from WebKit before coming to a final position, given the difference in behaviour around focus it seems worthwhile ensuring they're convinced by the design.

**Comment by janewman** [comment]
> Hi [@janewman](https://github.com/janewman), thanks for addressing those requests in the explainer.
> 
> We've discussed this further and would like to wait for a formal position from WebKit before coming to a final position, given the difference in behaviour around focus it seems worthwhile ensuring they're convinced by the design.

Wanted to point out this issue on the focus behavior for when user agents have different expectations for the default focusability of some native elements: https://github.com/openui/open-ui/issues/1306

It felt to me like there was consensus, and the behavior that Ryosuke advocated for was the one adopted, where focusgroup does not allow elements that would not appear in the [sequential focus navigation order](https://html.spec.whatwg.org/multipage/interaction.html#sequential-focus-navigation-order) to participate.
